main/lsp/request_dispatch.cc

```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
+                    return false;
+                } else if (hashes[i].definitions.hierarchyHash != core::GlobalStateHash::HASH_STATE_INVALID &&
+                           hashes[i].definitions.hierarchyHash != oldHash.definitions.hierarchyHash) {
+                    logger.debug("Taking slow path because {} has changed definitions", f->path());
+                    prodCategoryCounterInc("lsp.slow_path_reason", "changed_definition");
+                    return false;
+                }
+            }
+        }
+    }
+    logger.debug("Taking fast path");
+    return true;
+}
+
+UnorderedMap<int, core::FileHash> mergeEvictions(const UnorderedMap<int, core::FileHash> &olderEvictions,
+                                                 const UnorderedMap<int, core::FileHash> &newerEvictions) {
+    // For evictions, which are needed for emulating an older `globalStateHashes`, we keep the oldest.
+    UnorderedMap<int, core::FileHash> combinedEvictions = olderEvictions;
+    for (auto &e : newerEvictions) {
+        if (!combinedEvictions.contains(e.first)) {
+            combinedEvictions[e.first] = e.second;
+        }
+    }
+    return combinedEvictions;
+}
+
 } // namespace
 
+LSPFileUpdates LSPLoop::commitEdit(SorbetWorkspaceEditParams &edit) {
+    LSPFileUpdates update;
+    update.epoch = edit.epoch;
+    update.editCount = edit.mergeCount + 1;
+    update.updatedFileHashes = LSPTypechecker::computeFileHashes(*config, edit.updates, *emptyWorkers);
+    update.updatedFiles = move(edit.updates);
+    update.canTakeFastPath = canTakeFastPath(*initialGS, *config, globalStateHashes, update);
+    update.cancellationExpected = edit.sorbetCancellationExpected;
+
+    // Update globalStateHashes. Keep track of file IDs for these files, along with old hashes for these files.
+    vector<core::FileRef> frefs;
+    UnorderedMap<int, core::FileHash> evictedHashes;
+    {
+        ENFORCE(update.updatedFiles.size() == update.updatedFileHashes.size());
+        core::UnfreezeFileTable fileTableAccess(*initialGS);
+        int i = -1;
+        for (auto &file : update.updatedFiles) {
+            auto fref = initialGS->findFileByPath(file->path());
+            i++;
+            if (fref.exists()) {
+                ENFORCE(fref.id() < globalStateHashes.size());
+                initialGS = core::GlobalState::replaceFile(move(initialGS), fref, file);
+            } else {
+                // This file update adds a new file to GlobalState.
+                update.hasNewFiles = true;
+                fref = initialGS->enterFile(file);
+                fref.data(*initialGS).strictLevel = pipeline::decideStrictLevel(*initialGS, fref, config->opts);
+                if (fref.id() >= globalStateHashes.size()) {
+                    globalStateHashes.resize(fref.id() + 1);
+                }
+            }
+            evictedHashes[fref.id()] = move(globalStateHashes[fref.id()]);
```

Comment:
This is where we save file hashes that this edit evicts.

Reply:
Since `updatedFiles` is a vector, not a set, checking...is there a guarantee that its values are unique?  (Or more importantly, that the fref is unique?)

If not, it looks like a file that was repeated (or resolved to a repeated fref) would evict the `evictedHashes` value from the first time around...is that possible, and if so, is that problematic?

Reply:
Good question. `updatedFiles` should not contain any duplicated files. It's a vector for efficiency reasons, but is effectively treated like a set.

`SorbetWorkspaceEditParams` are created from four sources:

* didOpen messages, which concern one file.
* didChange messages, which concern one file.
* didClose messages, which concern one file.
* Watchman messages, which concern a set of files.

Each of these are converted into `SorbetWorkspaceEditParams` by `LSPPreprocessor`, and are merged with adjacent `SorbetWorkspaceEditParams` via `SorbetWorkspaceEditParams::merge`. That helper function ensures that `SorbetWorkspaceEditParams.updates` contains no duplicates during merging.

So, tl;dr: It shouldn't be possible. But it may be worth adding a few sanity checks here and there.